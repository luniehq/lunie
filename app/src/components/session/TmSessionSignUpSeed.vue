<template>
  <SessionFrame>
    <TmFormStruct :submit="onSubmit.bind(this)">
      <h2 class="session-title">Backup code</h2>
      <div>
        <div class="session-main bottom-indent reorder">
          <Steps
            v-if="!session.mobile"
            :steps="[`Create`, `Password`, `Backup`]"
            active-step="Backup"
          />
          <TmSeed :value="fieldSeed" />
          <TmFormGroup
            class="field-checkbox"
            :error="$v.fieldWarning.$error"
            field-id="sign-up-warning"
            field-label
          >
            <div class="field-checkbox-input">
              <label class="field-checkbox-label" for="sign-up-warning">
                <input
                  id="sign-up-warning"
                  v-model="fieldWarning"
                  v-focus
                  type="checkbox"
                />
                I understand that lost seeds cannot be recovered.</label
              >
            </div>
            <TmFormMsg
              v-if="$v.fieldWarning.$error && !$v.fieldWarning.required"
              name="Recovery confirmation"
              type="required"
            />
            <TmFormMsg v-if="error" type="custom" :msg="errorMessage" />
          </TmFormGroup>
        </div>
        <div class="session-footer">
          <TmBtn value="Create" />
        </div>
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { mapState, mapGetters } from "vuex"
import { sameAs } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmFormMsg from "common/TmFormMsg"
import SessionFrame from "common/SessionFrame"
import Steps from "../../ActionModal/components/Steps"
import TmSeed from "common/TmSeed"

export default {
  name: `session-sign-up`,
  components: {
    TmBtn,
    SessionFrame,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    Steps,
    TmSeed,
  },
  data: () => ({
    error: false,
    errorMessage: ``,
  }),
  computed: {
    ...mapState([`session`, `signup`]),
    ...mapGetters([`network`, `networkSlug`, `isExtension`, `currentNetwork`]),
    fieldSeed: {
      get() {
        return this.$store.state.signup.signUpSeed
      },
    },
    fieldWarning: {
      get() {
        return this.$store.state.signup.signUpWarning
      },
      set(value) {
        this.$store.commit(`updateField`, { field: `signUpWarning`, value })
      },
    },
  },
  mounted() {
    this.$store.dispatch(`createSeed`).then((seedPhrase) => {
      this.$store.commit(`updateField`, {
        field: `signUpSeed`,
        value: seedPhrase,
      })
    })
  },
  beforeDestroy: function () {
    this.$store.dispatch(`resetSignUpData`)
  },
  methods: {
    async onSubmit() {
      this.$v.$touch()
      if (this.$v.$error) return
      try {
        await this.$store.dispatch(`createKey`, {
          seedPhrase: this.signup.signUpSeed,
          password: this.signup.signUpPassword,
          name: this.signup.signUpName,
          HDPath: this.currentNetwork.defaultHDPath,
          curve: this.currentNetwork.defaultCurve,
          network: this.network,
        })
        if (this.isExtension) {
          this.$router.push(`/`)
        } else {
          this.$router.push({
            name: "portfolio",
            params: {
              networkId: this.networkSlug,
            },
          })
        }
      } catch (error) {
        this.error = true
        this.errorMessage = error.message
      }
    },
  },
  validations: () => ({
    fieldWarning: { required: sameAs(() => true) },
  }),
}
</script>
