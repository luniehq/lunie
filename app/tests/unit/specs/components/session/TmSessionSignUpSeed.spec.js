import { createLocalVue, shallowMount } from "@vue/test-utils"
import Vuelidate from "vuelidate"
import TmSessionSignUpSeed from "session/TmSessionSignUpSeed"

describe(`TmSessionSignUpSeed`, () => {
  const localVue = createLocalVue()
  localVue.directive(`focus`, () => {})
  localVue.use(Vuelidate)

  let wrapper, $store

  beforeEach(() => {
    $store = {
      getters: {
        network: "lunie-net",
        networkSlug: "lunie",
        isExtension: false,
        currentNetwork: {
          accountTypes: `["cosmosStandard"]`,
        },
      },
      state: {
        session: { insecureMode: true },
        signup: {
          signUpSeed: `asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf`,
          signUpWarning: false,
        },
      },
      commit: jest.fn(),
      dispatch: jest.fn().mockResolvedValue(""),
      mutations: {
        updateField: jest.fn(),
      },
    }

    wrapper = shallowMount(TmSessionSignUpSeed, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn(),
        },
      },
      stubs: [`router-link`],
    })
  })

  it("renders", () => {
    expect(wrapper.vm.fieldSeed).not.toBe(``)
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`validation should fail if fieldWarning is not checked`, async () => {
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$v.fieldWarning.$error).toBe(true)
  })

  it(`should commit updateField on fieldWarning field change`, async () => {
    wrapper.vm.fieldWarning = true
    await wrapper.vm.onSubmit()
    expect($store.commit).toHaveBeenCalledWith(`updateField`, {
      field: `signUpWarning`,
      value: true,
    })
  })

  it(`should dispatch createKey and route to / if form is submitted`, async () => {
    wrapper.vm.$store.state.signup.signUpSeed = `asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf`
    wrapper.vm.$store.state.signup.signUpPassword = `1234567890`
    wrapper.vm.$store.state.signup.signUpName = `HappyLunieUser`
    wrapper.vm.$store.state.signup.signUpWarning = true
    await wrapper.vm.onSubmit()
    expect($store.dispatch).toHaveBeenCalledWith(`createKey`, {
      name: "HappyLunieUser",
      password: "1234567890",
      network: "lunie-net",
      seedPhrase:
        "asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf",
      HDPath: undefined,
      curve: undefined,
    })
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith({
      name: "portfolio",
      params: { networkId: "lunie" },
    })
  })

  it(`should dispatch resetSignUpData when the component is destroyed`, async () => {
    wrapper.destroy()
    expect($store.dispatch).toHaveBeenCalledWith(`resetSignUpData`)
  })

  it(`should show error on createKey dispatch error`, async () => {
    wrapper.vm.$store.state.signup.signUpSeed = `asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf`
    wrapper.vm.$store.state.signup.signUpPassword = `1234567890`
    wrapper.vm.$store.state.signup.signUpName = `HappyLunieUser`
    wrapper.vm.$store.state.signup.signUpWarning = true
    wrapper.vm.$store.dispatch = {
      createKey: jest.fn().mockRejectedValue(new Error()),
    }
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.error).toBe(true)
  })
})
