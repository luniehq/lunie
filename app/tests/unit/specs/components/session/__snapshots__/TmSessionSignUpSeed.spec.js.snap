// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionSignUpSeed renders 1`] = `
<sessionframe-stub
  icon=""
>
  <tmformstruct-stub
    submit="function () { [native code] }"
    width=""
  >
    <h2
      class="session-title"
    >
      Backup code
    </h2>
     
    <div>
      <div
        class="session-main bottom-indent reorder"
      >
        <steps-stub
          activestep="Backup"
          steps="Create,Password,Backup"
        />
         
        <tmseed-stub
          value="asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf"
        />
         
        <tmformgroup-stub
          class="field-checkbox"
          fieldid="sign-up-warning"
          fieldlabel=""
        >
          <div
            class="field-checkbox-input"
          >
            <label
              class="field-checkbox-label"
              for="sign-up-warning"
            >
              <input
                id="sign-up-warning"
                type="checkbox"
              />
              
              I understand that lost seeds cannot be recovered.
            </label>
          </div>
           
          <!---->
           
          <!---->
        </tmformgroup-stub>
      </div>
       
      <div
        class="session-footer"
      >
        <tmbtn-stub
          customclass=""
          value="Create"
        />
      </div>
    </div>
  </tmformstruct-stub>
</sessionframe-stub>
`;
